### ISSUE TITLE
WFP/Firewall ( mcafee ) Idle connection issues

### ISSUE THREAD

**Issue by zzzjim** [basil00/Divert#206]
When running any WinDivert application idle connections have issues with McAfee.

To reproduce:

start filtering
```
passthru.exe ip
```

open a connection anywhere any kind ( tcp ) that idles
bbc sample

```
openssl s_client -host 151.101.0.81 -port 443
[PASTE]
Get / HTTP/1.1
Host: www.bbc.co.uk
Connection: Keep-Alive
[ENTER]
``` 

Wait 30 seconds
```
[PASTE]
Get / HTTP/1.1
Host: www.bbc.co.uk
Connection: Keep-Alive
[ENTER]
```

No response arrives on the second request. However, many packets are re-transmitted but none ever arrive to the application ( openssl ).

It seems the re-injected packets are possibly not seen by the mcafee firewall so the state of the connection goes into drop mode when the response arrives.

If pass passthru is not running the connection continues to function normally as expected.

Seems to be related to the tcp timeout in mcafee products/firewall.
But it doesn't happen if packets are not running through divert.
https://kc.mcafee.com/corporate/index?page=content&id=KB84567

This has been tested using the consumer version of McAfee. Default trial install.

Tested with: WinDivert-2.1.0-A
The problem is the same with  WinDivert1.X

I've tried filtering out impostor packets with no change

I've tried changing the priority to very high and very low with no change.

[mfe_pcaps_wfp_state.zip](https://github.com/basil00/Divert/files/3612136/mfe_pcaps_wfp_state.zip)

**Comment by basil00** [comment]
> It seems the re-injected packets are possibly not seen by the mcafee firewall so the state of the connection goes into drop mode when the response arrives.

Is this just a problem with mcafee firewall and not WinDivert?  Also, afaik, there should be no reason why WinDivert reinjected packets would be not visible to other drivers/firewalls.

**Comment by zzzjim** [comment]
It is only a problem when WinDivert is running.  If I run the same tests without WinDivert, McAfee firewall doesn't cause any issues.

**Comment by zzzjim** [comment]
I tried the same test with the inspect sample. It doesn't have the problem.

Possibly it is because each packet in WinDivert gets send as a newly allocated packet when it is written back to the driver.

The inspect sample clones the original packet rather then allocating it.

Windivert: FwpsAllocateNetBufferAndNetBufferList0
inpect:  FwpsAllocateCloneNetBufferList

Maybe this is why other WFP drivers are having issues with WinDivert

I'm guessing a packet id/ref sent to user space with the original packet memory would be needed and clones of all the queued packets maintained in the driver.

I haven't checked but maybe the following doesn't work when they are not cloned either.
      if ((packetState == FWPS_PACKET_INJECTED_BY_SELF) ||
          (packetState == FWPS_PACKET_PREVIOUSLY_INJECTED_BY_SELF))


WINDIVERT_ADDRESS {
...
   packet_id
}
WinDivertRecv(buf, address)
WindivertSetVerdict(... buf, address, DROP|ALLOW)
    - Inject the clone of the original packet

**Comment by basil00** [comment]
Sorry for late reply.

What happens if you append the string `&& not impostor` to the WinDivert filter?

> If I run the same tests without WinDivert, McAfee firewall doesn't cause any issues.

This can indicate a problem with the McAfee driver, e.g., if it does not handle WinDivert-injected packets properly.  It is usually a good idea to also report the issue to them as well.

> I'm guessing a packet id/ref sent to user space with the original packet memory would be needed and clones of all the queued packets maintained in the driver.

WFP drivers are not supposed to keep clones indefinitely, which means the driver cannot keep a clone while waiting for a user application verdict (which may never come).

Another problem is that cloning and reading packet contents outside of the classify function can cause some other buggy-but-popular drivers to crash.

> I haven't checked but maybe the following doesn't work when they are not cloned either.

It should work otherwise all injected packets will enter an infinite loop.

**Comment by zzzjim** [comment]
Hi,

Thanks for responding.

I did try adding "!impostor" to filter.  But it still behaves the same.

passthru.exe "ip && not impostor"

I was thinking of keeping the clone either until user space responds, a timeout occurs or the user space context is destroyed. More along the line of linux/ipq https://linux.die.net/man/3/libipq


> Another problem is that cloning and reading packet contents outside of the classify function can cause some other buggy-but-popular drivers to crash.
What would the difference be from accessing the data in "windivert_worker"?


I just noticed another difference.  The use of UNSPECIFIED_COMPARTMENT_ID in FwpsInjectForwardAsync0/FwpsInjectNetworkReceiveAsync0. The inspect sample saves the compartment id for re-injecting.

I'll let you know if I come up with anything else.

**Comment by basil00** [comment]
I think your best bet is to report to mcafee.  The symptoms of this bug can be explained by a bug in their driver, such as mishandling `FWPS_PACKET_INJECTED_BY_OTHER`.

> I was thinking of keeping the clone either until user space responds, a timeout occurs or the user space context is destroyed.

IIRC the clone still references the memory buffer belonging to the sending application.  So, the application will block until the clone is destroyed.

This can affect performance.  In fact, some earlier versions of WinDivert would queue cloned packets but the performance was worse than deep copying.

> The inspect sample saves the compartment id for re-injecting.

Yes, but afaik the compartment ID isn't actually used for anything.  As far as I can tell, it was intended for a feature that was eventually cut from Windows but the ID remains as an artifact.

**Comment by zzzjim** [comment]
I just tried passing the original compartment id back and forth and using in on re-injection.

passthru.exe ip 1 1

Before:
read packet_id: 11 Compartment: 0
read packet_id: 12 Compartment: 1
read packet_id: 13 Compartment: 1
read packet_id: 15 Compartment: 0
read packet_id: 16 Compartment: 0
read packet_id: 17 Compartment: 1
read packet_id: 18 Compartment: 0
read packet_id: 19 Compartment: 1

After:( All 1's)
read packet_id: 1 Compartment: 1
read packet_id: 2 Compartment: 1
read packet_id: 3 Compartment: 1
read packet_id: 4 Compartment: 1
read packet_id: 5 Compartment: 1
read packet_id: 6 Compartment: 1
read packet_id: 7 Compartment: 1

It didn't fix the issue, but worth a try.

**Comment by zzzjim** [comment]
Tried saving the original packets and it still didn't solve the issue.

Tried adding NdisCopySendNetBufferListInfo/NdisCopyReceiveNetBufferListInfo which also didn't work.

Tried using different handles for injecting inbound vs outbound. Problem solved. I'm not sure why this makes a difference but it works. Found this based on WFPSampler as it does't have the issue.

Add a pull request #211  with the changes.

Will you still be making builds of the 1.x series?  The current 2.x doesn't load on windows 7 with the new driver signing.

**Comment by basil00** [comment]
Thanks for finding a solution to this problem.

> Found this based on WFPSampler as it doesn't have the issue.

This is a convincing argument for having multiple handles.  Still, not sure why it makes a difference and perhaps it is undocumented.

> Will you still be making builds of the 1.x series? The current 2.x doesn't load on windows 7 with the new driver signing.

I will not be updating 1.X anymore.  The 2.X loads on Windows 7 after update or at least if  KB3033929 is installed.
